Mathlib/Algebra/Homology/ShortComplex/Basic.lean

```diff
@@ -0,0 +1,294 @@
+  obj S := S.X₃
+  map f := f.τ₃
+
+instance preservesZeroMorphisms_π₁ : Functor.PreservesZeroMorphisms (π₁ : _ ⥤ C) where
+instance preservesZeroMorphisms_π₂ : Functor.PreservesZeroMorphisms (π₂ : _ ⥤ C) where
+instance preservesZeroMorphisms_π₃ : Functor.PreservesZeroMorphisms (π₃ : _ ⥤ C) where
+
+instance (f : S₁ ⟶ S₂) [IsIso f] : IsIso f.τ₁ := (inferInstance : IsIso (π₁.mapIso (asIso f)).hom)
+instance (f : S₁ ⟶ S₂) [IsIso f] : IsIso f.τ₂ := (inferInstance : IsIso (π₂.mapIso (asIso f)).hom)
+instance (f : S₁ ⟶ S₂) [IsIso f] : IsIso f.τ₃ := (inferInstance : IsIso (π₃.mapIso (asIso f)).hom)
+
+/-- The natural transformation `π₁ ⟶ π₂` induced by `S.f` for all `S : ShortComplex C`. -/
+@[simps] def π₁Toπ₂ : (π₁ : _ ⥤ C) ⟶ π₂ where
+  app S := S.f
+
+/-- The natural transformation `π₂ ⟶ π₃` induced by `S.g` for all `S : ShortComplex C`. -/
+@[simps] def π₂Toπ₃ : (π₂ : _ ⥤ C) ⟶ π₃ where
+  app S := S.g
+
+@[reassoc (attr := simp)]
+lemma π₁Toπ₂_comp_π₂Toπ₃ : (π₁Toπ₂ : (_ : _ ⥤ C) ⟶ _) ≫ π₂Toπ₃ = 0 := by aesop_cat
+
+variable {D}
+variable [HasZeroMorphisms D]
+
+/-- The short complex in `D` obtained by applying a functor `F : C ⥤ D` to a
+short complex in `C`, assuming that `F` preserves zero morphisms. -/
+@[simps]
+def map (F : C ⥤ D) [F.PreservesZeroMorphisms] : ShortComplex D :=
+  ShortComplex.mk (F.map S.f) (F.map S.g) (by rw [← F.map_comp, S.zero, F.map_zero])
+
+/-- The morphism of short complexes `S.map F ⟶ S.map G` induced by
+a natural transformation `F ⟶ G`. -/
+@[simps]
+def mapNatTrans {F G : C ⥤ D} [F.PreservesZeroMorphisms] [G.PreservesZeroMorphisms] (τ : F ⟶ G) :
+    S.map F ⟶ S.map G where
+  τ₁ := τ.app _
+  τ₂ := τ.app _
+  τ₃ := τ.app _
+
+/-- The isomorphism of short complexes `S.map F ≅ S.map G` induced by
+a natural isomorphism `F ≅ G`. -/
+@[simps]
+def mapNatIso {F G : C ⥤ D} [F.PreservesZeroMorphisms] [G.PreservesZeroMorphisms] (τ : F ≅ G) :
+    S.map F ≅ S.map G where
+  hom := S.mapNatTrans τ.hom
+  inv := S.mapNatTrans τ.inv
+
+/-- The functor `ShortComplex C ⥤ ShortComplex D` induced by a functor `C ⥤ D` which
+preserves zero morphisms. -/
+@[simps]
+def _root_.CategoryTheory.Functor.mapShortComplex (F : C ⥤ D) [F.PreservesZeroMorphisms] :
+    ShortComplex C ⥤ ShortComplex D where
+  obj S := S.map F
+  map φ :=
+    { τ₁ := F.map φ.τ₁
+      τ₂ := F.map φ.τ₂
+      τ₃ := F.map φ.τ₃
+      comm₁₂ := by
+        dsimp
+        simp only [← F.map_comp, φ.comm₁₂]
+      comm₂₃ := by
+        dsimp
+        simp only [← F.map_comp, φ.comm₂₃] }
+
+/-- A constructor for isomorphisms in the category `ShortComplex C`-/
+@[simps]
+def isoMk (e₁ : S₁.X₁ ≅ S₂.X₁) (e₂ : S₁.X₂ ≅ S₂.X₂) (e₃ : S₁.X₃ ≅ S₂.X₃)
+    (comm₁₂ : e₁.hom ≫ S₂.f = S₁.f ≫ e₂.hom) (comm₂₃ : e₂.hom ≫ S₂.g = S₁.g ≫ e₃.hom) :
+    S₁ ≅ S₂ where
+  hom := ⟨e₁.hom, e₂.hom, e₃.hom, comm₁₂, comm₂₃⟩
+  inv := homMk e₁.inv e₂.inv e₃.inv
+    (by rw [← cancel_mono e₂.hom, assoc, assoc, e₂.inv_hom_id, comp_id,
+      ← comm₁₂, e₁.inv_hom_id_assoc])
```

Comment:
I don't know if the spacing is correct here. Should this be under the `rw` rather than under the `by`? Not sure.